Issue4223 add banner after case contact creation

vasconsaurus opened the pull request:
### What github issue is this PR for, if any?
Resolves #4223 

### What changed, and why?
After the user adds a new case contact to a case the modal displays a different message according to the court dates:

- if the case only has court dates in the **future**:
_Please enter the previous court date if you know it so that court reports can be generated with the correct case contacts._
- If the case only has court dates in the **past**:
_Please enter the next court date if you know it so that court reports can be generated with the correct case contacts._
- if the case has **both** court dates in the past and future **it doesn't display a message**

note: The issue asks for a banner, but after talking this through with Shen, we decided it would be better to use the modal to display this.

### How is this tested? (please write tests!) 💖💪
- spec/helpers/court_dates_helper_spec.rb

### Screenshots please :)
1. Both
<img width="1120" alt="Screenshot 2022-12-05 at 12 51 56" src="https://user-images.githubusercontent.com/87862340/205682891-960ec83d-e5a4-4fc6-a405-cc5260078c57.png">
<img width="1120" alt="Screenshot 2022-12-05 at 12 52 03" src="https://user-images.githubusercontent.com/87862340/205682939-bfcfc00d-3766-45a7-80e3-208731034599.png">

2. Future
<img width="1120" alt="Screenshot 2022-12-05 at 12 52 30" src="https://user-images.githubusercontent.com/87862340/205683090-8bd77872-426a-4266-b72b-a7b99fa95d17.png">
<img width="1120" alt="Screenshot 2022-12-05 at 12 52 21" src="https://user-images.githubusercontent.com/87862340/205683109-1a144235-e1b1-489f-97d6-4657bfa13ad6.png">

3. Past
<img width="1120" alt="Screenshot 2022-12-05 at 12 52 43" src="https://user-images.githubusercontent.com/87862340/205683171-a69e7dd6-f63f-41a3-91a5-543a814081e0.png">
<img width="1120" alt="Screenshot 2022-12-05 at 12 52 58" src="https://user-images.githubusercontent.com/87862340/205683210-94eb1764-298c-4584-9ad6-4216e8c01516.png">

### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
`![alt text](https://media.giphy.com/media/1nP7ThJFes5pgXKUNf/giphy.gif)`

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

vasconsaurus committed:
added a banner when a new case contact is created

vasconsaurus committed:
built the logic for our banner

vasconsaurus committed:
added the message to the modal

vasconsaurus committed:
Moved the logic to a court_dates_helper

Not to sure about it yet though

vasconsaurus committed:
reorganized the divs so an empty div isnt created

The way it was organized before, I would get an empty div
when there was no prompt. This way we only have that div if
there is a prompt

vasconsaurus committed:
Merge branch 'main' into issue4223-add_banner_after_case_contact_creation

vasconsaurus committed:
refactored a bit

vasconsaurus committed:
refactored to remove loop

If the last court date is the past, and the query returns them
in ascending order, then all are in the past. If the last is in
the future, then the first could be in the past, so we check for that.
So we only need to check the first and last court_dates, not all.

vasconsaurus committed:
created first version of court_dates_helper test

vasconsaurus committed:
linted

vasconsaurus committed:
added a test for when a case has dates in the future and past

tcione commented on app/helpers/court_dates_helper.rb:
```diff
@@ -0,0 +1,16 @@
+
+    if last_court_date.date < Date.current
+      "past"
+    elsif last_court_date.date > Date.current && first_court_date.date > Date.current
```
**suggestion:** it looks to me that you could get away with only `first_court_date.date > Date.current` instead of `last_court_date.date > Date.current && first_court_date.date > Date.current`. WDYT?

vasconsaurus replied:
derp! agree! 🙂

tcione commented on app/helpers/court_dates_helper.rb:
```diff
@@ -0,0 +1,16 @@
+# Helper methods for court_dates
+module CourtDatesHelper
+  def when_do_we_have_court_dates(casa_case)
+    date_query = casa_case.court_dates.includes(:hearing_type).ordered_ascending
```
**nitpick:** why `date_query` over `court_dates`?

**question:** do we really need to include `hearing_type`?

vasconsaurus replied:
> why date_query over `court_dates`?

Because it was querying for the date and I was a bit too literal, I agree `court_dates` is much better

> do we really need to include `hearing_type`?

I don't think so. Removed it 🙂

tcione commented on app/helpers/court_dates_helper.rb:
```diff
@@ -0,0 +1,16 @@
+module CourtDatesHelper
+  def when_do_we_have_court_dates(casa_case)
+    date_query = casa_case.court_dates.includes(:hearing_type).ordered_ascending
+    last_court_date = date_query.last
+    first_court_date = date_query.first
```
**suggestion:** it feels to me like you could use `date_query.last.date` directly instead of storing this in a variable. It also seems to me like it doesn't lose in meaning/how-understandable-it-is. WDYT?

tcione commented on app/helpers/court_dates_helper.rb:
```diff
@@ -0,0 +1,16 @@
+    last_court_date = date_query.last
+    first_court_date = date_query.first
+
+    if last_court_date.date < Date.current
```
**suggestion:** it shouldn't make a difference (unless a user is unlucky enough for this script to be executed at the exact moment the code is running), but if you want to ensure we're always using the date date, it'd be better to store `Date.current` in a variable and use it instead.

tcione commented on app/views/casa_cases/_thank_you_modal.html.erb:
```diff
@@ -2,9 +2,18 @@
+        <div class="modal-header">
           <%= "Case contact was successfully created. #{thank_you_message}" %>
         </div>
+        <% if when_do_we_have_court_dates(@casa_case) == 'future' %>
```
**nitpick:** you are using `"future"` in the module and `'future'` here. I tend to think being consistent with quotes is better. That said, that's obviously me being annoying, so up to you listening to this or not.

tcione commented on app/views/casa_cases/_thank_you_modal.html.erb:
```diff
@@ -2,9 +2,18 @@
+          <div id="prompt-update-user" class="modal-body">
+            Please enter the previous court date if you know it so that court reports can be generated with the correct case contacts.
+          </div>
+        <% elsif when_do_we_have_court_dates(@casa_case) == 'past' %>
```
**suggestion:** store `when_do_we_have_court_dates(@casa_case)` in a variable to avoid querying the database multiple times at different points in time.

tcione commented on spec/factories/casa_cases.rb:
```diff
@@ -68,4 +68,10 @@
     end
   end
+
+  trait :with_past_court_date do
```
**praise:** :pinched_fingers: :kiss:

tcione commented on spec/helpers/court_dates_helper_spec.rb:
```diff
@@ -0,0 +1,29 @@
+    describe "when casa case has only dates in the past" do
+      let(:casa_case) { create(:casa_case, :with_past_court_date) }
+
+      it "returns 'past'" do
+        expect(helper.when_do_we_have_court_dates(casa_case)).to eq("past")
+      end
```
**nitpick:** I tend to think these types of tests a bit too repetitive, the case description adds very little to the actual code being executed. In these cases I prefer something terser:

```rb
it { expect(helper.when_do_we_have_court_dates(casa_case)).to eq("past") }
```

That said, up to taste :)

tcione commented on spec/helpers/court_dates_helper_spec.rb:
```diff
@@ -0,0 +1,29 @@
+      let(:casa_case) { create(:casa_case, :with_upcoming_court_date) }
+
+      it "returns 'future'" do
+        expect(helper.when_do_we_have_court_dates(casa_case)).to eq("future")
```
**suggestion:** make `helper.when_do_we_have_court_dates(casa_case)` `describe "#when_do_we_have_court_dates" do`'s subject

tcione commented on spec/helpers/court_dates_helper_spec.rb:
```diff
@@ -0,0 +1,29 @@
+      let(:casa_case) { create(:casa_case, :with_upcoming_court_date, :with_past_court_date) }
+
+      it "returns nothing" do
+        expect(helper.when_do_we_have_court_dates(casa_case)).to eq(nil)
```
**nitpick:**
```suggestion
        expect(helper.when_do_we_have_court_dates(casa_case)).to be_nil
```

tcione commented on spec/helpers/court_dates_helper_spec.rb:
```diff
@@ -0,0 +1,29 @@
+require "rails_helper"
+
+RSpec.describe CourtDatesHelper do
```
**praise:** always nice to see tests for helpers :D

tcione reviewed (commented):
Nice work there, I like how thoughtful is the description and the approach takes. That said I have a few smallish comments for your consideration.

vasconsaurus committed:
updated logic

– If the first is the future, then the last is also
– Updated court_dates query, I don't think we need to include
hearing_type in this one
– stored when_do_we_have_court_dates(@casa_case) in a variable
to avoid querying the database multiple times at different points
in time.

vasconsaurus committed:
rewrote the helper test to be less repetitive

– It's a simple test, it's shorter but still clear
– Added a subject to remove repetition
– Added returns if case or dates are nil

compwron commented on app/helpers/court_dates_helper.rb:
```diff
@@ -0,0 +1,19 @@
+
+# Helper methods for court_dates
+module CourtDatesHelper
+  def when_do_we_have_court_dates(casa_case)
```
we should look at naming here

compwron reviewed (approved):
:)
